### PR TITLE
Distinguish available liquidity per swap amm

### LINKED ISSUE #111: ODHack: Distinguish available liquidity per swap AMM
Distinguish parts of the available liquidity coming from different swap AMMs.

Steps:
1) Check out the current implementation of fetching available liquidity from Uniswap V2 AMMs: [https://github.com/CarmineOptions/derisk-research/blob/master/src/swap_amm.py#L369](https://github.com/CarmineOptions/derisk-research/blob/master/src/swap_amm.py#L369).
2) Currently, the liquidity from all AMMs is mixed together. Adjust the code so that we're able to distinguish between the liquidity available, e.g., on MySwap, SithSwap, JediSwap, or 10kSwap.
3) Save the adjusted data to a new file on the Google Storage.
4) Visualize available liquidity from different swap AMMs on the main chart on the dashboard. Run `make app` to validate the changes locally.

### PR DESCRIPTION
**Pull request by adrianvrj**
Closes #111 

Current output

telegram-cloud-photo-size-1-5100694837539220366-y

### CODE REVIEW

File: src/main_chart.py [hunk 1]
```diff
@@ -23,34 +20,40 @@ def get_main_chart_data(
...
+
+    # Initialize columns for each AMM
+    for amm in ["jedi", "10k", "my", "sith"]:
+        data[f"{amm}_debt_token_supply"] = 0
+
+    for index, row in data.iterrows():
+        price = row["collateral_token_price"]
+        total_supply = 0
+        for amm in ["jedi", "10k", "my", "sith"]:
+            supply = swap_amms.get_supply_at_price(collateral_token, price, debt_token, amm)
```
Comment (lukaspetrasek) [inline src/main_chart.py:47]: Add variable names to the call to `get_supply_at_price`.

File: src/main_chart.py [hunk 2]
```diff
@@ -23,34 +20,40 @@ def get_main_chart_data(
...
+            data.at[index, f"{amm}_debt_token_supply"] = supply
```
Comment (lukaspetrasek) [inline src/main_chart.py:48]: Use `data['collateral_token_price'].apply` instead of the for loop.

File: src/main_chart.py [hunk 3]
```diff
@@ -23,34 +20,40 @@ def get_main_chart_data(
...
+            total_supply += supply
+        data.at[index, "debt_token_supply"] = total_supply
```
Comment (lukaspetrasek) [inline src/main_chart.py:50]: This can be the sum over the 4 AMM columns.

File: src/main_chart.py [hunk 4]
```diff
@@ -59,22 +62,26 @@ def get_main_chart_figure(
     collateral_token: str,
     debt_token: str,
 ) -> plotly.graph_objs.Figure:
+    # Define the AMMs and their color mappings
+    amms = ["jedi", "10k", "my", "sith"]
+    color_map = {"jedi": "#ECD662", "10k": "#4CA7D0", "my": "#76C1FA", "sith": "#FFA07A"}
```
Comment (lukaspetrasek) [inline src/main_chart.py:67]: Use different shades of blue, to be consistent with the current version of the chart.

File: src/main_chart.py [hunk 5]
```diff
@@ -59,22 +62,26 @@ def get_main_chart_figure(
     collateral_token: str,
     debt_token: str,
 ) -> plotly.graph_objs.Figure:
+    # Define the AMMs and their color mappings
+    amms = ["jedi", "10k", "my", "sith"]
```
Comment (lukaspetrasek) [inline src/main_chart.py:66]: This could be a constant in the `swap_amm.py` module.

File: src/main_chart.py [hunk 6]
```diff
@@ -59,22 +62,26 @@ def get_main_chart_figure(
...
-        y=["liquidable_debt_at_interval", "debt_token_supply"],
-        title=f"Liquidable debt and the corresponding supply of {debt_token} at various price intervals of "
-            f"{collateral_token}",
+        y=[f"{amm}_debt_token_supply" for amm in amms] + ["liquidable_debt_at_interval"],
+        title=f"Liquidable debt and the corresponding supply of {debt_token} at various price intervals of {collateral_token}",
         barmode="overlay",
         opacity=0.65,
-        # TODO: Align colors with the rest of the app.
-        color_discrete_map={"liquidable_debt_at_interval": "#ECD662", "debt_token_supply": "#4CA7D0"},
+        color_discrete_map={**color_map, "liquidable_debt_at_interval": "#FFD700"},
```
Comment (lukaspetrasek) [inline src/main_chart.py:76]: Pls keep the TODO comment in place.

File: src/swap_amm.py [hunk 7]
```diff
@@ -322,14 +324,18 @@ def __init__(self, symbol1, symbol2, addresses, myswap_id):
...
             myswap_pool = await src.blockchain_call.get_myswap_pool(self.myswap_id)
-        for token in self.tokens:
-            balance = 0
-            for address in self.addresses:
-                balance += await src.blockchain_call.balance_of(token.address, address)
-            if self.myswap_id is not None:
-                balance += myswap_pool[token.symbol.upper()]
-            token.balance_base = balance
-            token.balance_converted = decimal.Decimal(balance) / token.decimal_factor
+        for amm, addresses in self.addresses_dict.items():
```
Comment (lukaspetrasek) [inline src/swap_amm.py:327]: `addresses` is always a single address, right?
```suggestion
        for amm, address in self.addresses_dict.items():
```

File: src/swap_amm.py [hunk 8]
```diff
@@ -322,14 +324,18 @@ def __init__(self, symbol1, symbol2, addresses, myswap_id):
...
+            for token in self.tokens:
+                balance = 0
+                if isinstance(addresses, str):
+                    addresses = [addresses]
+                for address in addresses:
+                    balance += await src.blockchain_call.balance_of(token.address, address)
+                if self.myswap_id is not None:
+                    balance += myswap_pool[token.symbol.upper()]
```
Comment (lukaspetrasek) [inline src/swap_amm.py:335]: This always adds the `MySwap` balance even if `amm` is not `MySwap`.

File: src/swap_amm.py [hunk 9]
```diff
@@ -322,14 +324,18 @@ def __init__(self, symbol1, symbol2, addresses, myswap_id):
...
+                self.balances[amm][token.symbol] = balance
+                token.balance_base = sum(self.balances[amm][token.symbol] for amm in self.balances)
```
Comment (lukaspetrasek) [inline src/swap_amm.py:337]: It's not necessary to perform this in every iteration.

**Comment by lukaspetrasek** [comment]
> Closes #111
> 
> Current output
> 
> telegram-cloud-photo-size-1-5100694837539220366-y

Looking at the chart, the values are too large. Each bar should roughly correspond to tens of thousands of dollars, so the scale of 10^15 is too high. Please investigate whether you're working with the decimals correctly, or whether you're not adding the liquidity multiple times.

**Review by lukaspetrasek — CHANGES_REQUESTED** [review]

File: src/swap_amm.py [hunk 10]
```diff
@@ -11,6 +10,8 @@
 
 
 
+AMMS = ["JediSwap", "10kSwap", "MySwap", "SithSwap"]
+
```
Comment (lukaspetrasek) [inline src/swap_amm.py:14]: Add two more blank lines 🙏🏼

File: src/main_chart.py [hunk 11]
```diff
@@ -59,22 +62,26 @@ def get_main_chart_figure(
     collateral_token: str,
     debt_token: str,
 ) -> plotly.graph_objs.Figure:
+    # Define the AMMs and their color mappings
+    amms = ["jedi", "10k", "my", "sith"]
+    color_map = {"jedi": "#ECD662", "10k": "#4CA7D0", "my": "#76C1FA", "sith": "#FFA07A"}
+
     figure = plotly.express.bar(
```
[the remark this answers is not in the document]
Comment (lukaspetrasek) [reply src/main_chart.py:73]: I'm still getting this error.

File: src/swap_amm.py (hunk 10 above)
Comment (adrianvrj) [reply src/swap_amm.py:14]: No problem! :)

File: src/main_chart.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (adrianvrj) [reply src/main_chart.py:73]: I'll take a closer look at that

File: src/swap_amm.py [hunk 12]
```diff
@@ -324,25 +325,21 @@ def __init__(self, symbol1, symbol2, addresses_dict, myswap_id):
...
-                token.balance_converted = decimal.Decimal(token.balance_base) / token.decimal_factor
+                token.balance_base = balance
+                token.balance_converted = balance
 
     def update_converted_balance(self):
         for token in self.tokens:
-            token.balance_converted = decimal.Decimal(token.balance_base) / token.decimal_factor
+            token.balance_converted = token.balance_base
 
     def buy_tokens(self, symbol, amount):
```
Comment (lukaspetrasek) [inline src/swap_amm.py:342]: Pls add typing.

File: src/main_chart.py [hunk 13]
```diff
@@ -1,16 +1,13 @@
 import decimal
-
-import pandas
-import plotly.express
+import pandas
```
[the remark this answers is not in the document]
Comment (lukaspetrasek) [reply src/main_chart.py:2]: Keeping unresolved.

File: src/main_chart.py (hunk 1 above)
Comment (lukaspetrasek) [reply src/main_chart.py:47]: Keeping unresolved.

File: src/swap_amm.py [hunk 14]
```diff
@@ -324,25 +325,21 @@ def __init__(self, symbol1, symbol2, addresses_dict, myswap_id):
...
             for token in self.tokens:
-                balance = 0
-                if isinstance(addresses, str):
-                    addresses = [addresses]
-                for address in addresses:
-                    balance += await src.blockchain_call.balance_of(token.address, address)
-                if self.myswap_id is not None:
+                balance = await src.blockchain_call.balance_of(token.address, address)
+                if self.myswap_id is not None and amm == "MySwap":
                     balance += myswap_pool[token.symbol.upper()]
```
Comment (lukaspetrasek) [inline src/swap_amm.py:332]: Now MySwap is still added twice, right? First on line 330, and then on line 332.

File: src/swap_amm.py [hunk 15]
```diff
@@ -324,25 +325,21 @@ def __init__(self, symbol1, symbol2, addresses_dict, myswap_id):
...
+                balance = decimal.Decimal(balance) / token.decimal_factor
                 self.balances[amm][token.symbol] = balance
-                token.balance_base = sum(self.balances[amm][token.symbol] for amm in self.balances)
-                token.balance_converted = decimal.Decimal(token.balance_base) / token.decimal_factor
+                token.balance_base = balance
+                token.balance_converted = balance
```
Comment (lukaspetrasek) [inline src/swap_amm.py:336]: Remove one of the field for balance then. And name the remaining one just `balance`.

Comment (lukaspetrasek) [inline src/swap_amm.py:336]: This gets overwritten at each iteration, so only the value for the last amm survives.

**Review by lukaspetrasek — CHANGES_REQUESTED** [review]

**Comment by lukaspetrasek** [comment]
Merged 🙏🏼

**Pull request merged**